File: src/main/java/com/bank/service/internal/DefaultTransferService.java
package com.bank.service.internal;

import com.bank.domain.Account;
import com.bank.domain.InsufficientFundsException;
import com.bank.domain.InvalidTimePolicy;
import com.bank.domain.TransferReceipt;
import com.bank.repository.AccountRepository;
import com.bank.service.FeePolicy;
import com.bank.service.TransferService;
import org.springframework.transaction.annotation.Transactional;

import static java.lang.String.format;

public class DefaultTransferService implements TransferService {

    private final AccountRepository accountRepository;
    private final FeePolicy feePolicy;
    private double minimumTransferAmount = 1.00;
    private TimePolicy timePolicy;

    public DefaultTransferService(AccountRepository accountRepository, FeePolicy feePolicy, TimePolicy timePolicy) {
        this.accountRepository = accountRepository;
        this.feePolicy = feePolicy;
        this.timePolicy = timePolicy;
    }

    @Override
    public void setMinimumTransferAmount(double minimumTransferAmount) {
        this.minimumTransferAmount = minimumTransferAmount;
    }

    @Override
    @Transactional
    public TransferReceipt transfer(double amount, String srcAcctId, String dstAcctId) throws InsufficientFundsException {
        if (amount < minimumTransferAmount) {
            throw new IllegalArgumentException(format("transfer amount must be at least $%.2f", minimumTransferAmount));
        }
        if (!timePolicy.isTimeValid()) {
            throw new InvalidTimePolicy("Not allow transfer at 22:00 - 05:59");
        }
        TransferReceipt receipt = new TransferReceipt();

        Account srcAcct = accountRepository.findById(srcAcctId);
        Account dstAcct = accountRepository.findById(dstAcctId);

        receipt.setInitialSourceAccount(srcAcct);
        receipt.setInitialDestinationAccount(dstAcct);

        double fee = feePolicy.calculateFee(amount);
        if (fee > 0) {
            srcAcct.debit(fee);
        }

        receipt.setTransferAmount(amount);
        receipt.setFeeAmount(fee);

        srcAcct.debit(amount);
        dstAcct.credit(amount);

        accountRepository.updateBalance(srcAcct);
        accountRepository.updateBalance(dstAcct);

        receipt.setFinalSourceAccount(srcAcct);
        receipt.setFinalDestinationAccount(dstAcct);

        return receipt;
    }
}

File: src/main/java/com/bank/service/internal/DefaultTransferService.kt
package com.bank.service.internal

import com.bank.domain.InsufficientFundsException
import com.bank.domain.InvalidTimePolicy
import com.bank.domain.TransferReceipt
import com.bank.repository.AccountRepository
import com.bank.service.FeePolicy
import com.bank.service.TransferService
import org.springframework.transaction.annotation.Transactional
import java.lang.String.format

open class DefaultTransferService(private val accountRepository: AccountRepository, private val feePolicy: FeePolicy, private val timePolicy: TimePolicy) : TransferService {
    private var minimumTransferAmount = 1.00

    override fun setMinimumTransferAmount(minimumTransferAmount: Double) {
        this.minimumTransferAmount = minimumTransferAmount
    }

    @Transactional
    @Throws(InsufficientFundsException::class)
    override fun transfer(amount: Double, srcAcctId: String, dstAcctId: String): TransferReceipt {
        if (amount < minimumTransferAmount) {
            throw IllegalArgumentException(format("transfer amount must be at least $%.2f", minimumTransferAmount))
        }
        if (timePolicy.isTimeValid().not()) {
            throw InvalidTimePolicy("Not allow transfer at 22:00 - 05:59")
        }
        val receipt = TransferReceipt()

        val srcAcct = accountRepository.findById(srcAcctId)
        val dstAcct = accountRepository.findById(dstAcctId)

        receipt.setInitialSourceAccount(srcAcct)
        receipt.setInitialDestinationAccount(dstAcct)

        val fee = feePolicy.calculateFee(amount)
        if (fee > 0) {
            srcAcct.debit(fee)
        }

        receipt.transferAmount = amount
        receipt.feeAmount = fee

        srcAcct.debit(amount)
        dstAcct.credit(amount)

        accountRepository.updateBalance(srcAcct)
        accountRepository.updateBalance(dstAcct)

        receipt.finalSourceAccount = srcAcct
        receipt.finalDestinationAccount = dstAcct

        return receipt
    }
}
